

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Using ParmEd with MDAnalysis and OpenMM to simulate a selection of atoms &mdash; MDAnalysis User Guide  documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/mdanalysis-logo.ico"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
        <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Alignments and RMS fitting" href="analysis/alignment_and_rms/README.html" />
    <link rel="prev" title="Centering a trajectory in the box" href="transformations/center_protein_in_box.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: white" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/user_guide.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Getting started</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick start guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">Frequently asked questions</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="README.html">Examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="constructing_universe.html">Constructing, modifying, and adding to a Universe</a></li>
<li class="toctree-l2"><a class="reference internal" href="transformations/README.html">Transformations</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Using ParmEd with MDAnalysis and OpenMM to simulate a selection of atoms</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Loading-files:-the-difference-between-ParmEd-and-MDAnalysis">Loading files: the difference between ParmEd and MDAnalysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Using-MDAnalysis-to-select-atoms">Using MDAnalysis to select atoms</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Using-ParmEd-and-OpenMM-to-create-a-simulation-system">Using ParmEd and OpenMM to create a simulation system</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="analysis/alignment_and_rms/README.html">Alignments and RMS fitting</a></li>
<li class="toctree-l2"><a class="reference internal" href="analysis/distances_and_contacts/README.html">Distances and contacts</a></li>
<li class="toctree-l2"><a class="reference internal" href="analysis/trajectory_similarity/README.html">Trajectory similarity</a></li>
<li class="toctree-l2"><a class="reference internal" href="analysis/hydrogen_bonding/README.html">Hydrogen bonding</a></li>
<li class="toctree-l2"><a class="reference internal" href="analysis/structure/README.html">Structure</a></li>
<li class="toctree-l2"><a class="reference internal" href="analysis/reduced_dimensions/README.html">Dimension reduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="analysis/polymers_and_membranes/README.html">Polymers and membranes</a></li>
<li class="toctree-l2"><a class="reference internal" href="analysis/volumetric/README.html">Volumetric analyses</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Data structures</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../universe.html">Universe</a></li>
<li class="toctree-l1"><a class="reference internal" href="../atomgroup.html">AtomGroup</a></li>
<li class="toctree-l1"><a class="reference internal" href="../groups_of_atoms.html">Groups of atoms</a></li>
<li class="toctree-l1"><a class="reference internal" href="../selections.html">Atom selection language</a></li>
<li class="toctree-l1"><a class="reference internal" href="../topology_system.html">The topology system</a></li>
</ul>
<p class="caption"><span class="caption-text">Trajectories</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/trajectories.html">Trajectories</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/slicing_trajectories.html">Slicing trajectories</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/transformations.html">On-the-fly transformations</a></li>
</ul>
<p class="caption"><span class="caption-text">Input/output</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../reading_and_writing.html">Reading and writing files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../formats/index.html">Format overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../formats/guessing.html">Guessing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../formats/auxiliary.html">Auxiliary files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../formats/selection_exporters.html">Selection exporters</a></li>
<li class="toctree-l1"><a class="reference internal" href="../formats/format_reference.html">Format reference</a></li>
</ul>
<p class="caption"><span class="caption-text">Analysis</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="analysis/README.html">Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis/README.html#alignments-and-rms-fitting">Alignments and RMS fitting</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis/README.html#distances-and-contacts">Distances and contacts</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis/README.html#trajectory-similarity">Trajectory similarity</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis/README.html#hydrogen-bonding">Hydrogen bonding</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis/README.html#structure">Structure</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis/README.html#volumetric-analyses">Volumetric analyses</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis/README.html#dimension-reduction">Dimension reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis/README.html#polymers-and-membranes">Polymers and membranes</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis/custom_trajectory_analysis.html">Writing your own trajectory analysis</a></li>
</ul>
<p class="caption"><span class="caption-text">Advanced</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../standard_selections.html">Standard residues in MDAnalysis selections</a></li>
<li class="toctree-l1"><a class="reference internal" href="../advanced_topology.html">Advanced topology concepts</a></li>
<li class="toctree-l1"><a class="reference internal" href="../datasets.html">Example data</a></li>
</ul>
<p class="caption"><span class="caption-text">Contributing</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributing to MDAnalysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing_code.html">Contributing to the main codebase</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing_docs.html">Contributing to the user guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preparing_releases_and_hotfixes.html">Preparing a release</a></li>
<li class="toctree-l1"><a class="reference internal" href="../module_imports.html">Module imports in MDAnalysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../testing.html">Tests in MDAnalysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MDAnalysis User Guide</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="README.html">Examples</a> &raquo;</li>
        
      <li>Using ParmEd with MDAnalysis and OpenMM to simulate a selection of atoms</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/examples/parmed_sim.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    min-width: 5ex;
    padding-top: 0.3rem;
    padding-right: 0.3rem;
    text-align: right;
    flex: 0;
}
@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 0.3rem;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<script src='http://cdnjs.cloudflare.com/ajax/libs/require.js/2.1.10/require.min.js'></script>
<script>require=requirejs;</script><div class="section" id="Using-ParmEd-with-MDAnalysis-and-OpenMM-to-simulate-a-selection-of-atoms">
<h1>Using ParmEd with MDAnalysis and OpenMM to simulate a selection of atoms<a class="headerlink" href="#Using-ParmEd-with-MDAnalysis-and-OpenMM-to-simulate-a-selection-of-atoms" title="Permalink to this headline">¶</a></h1>
<p>Here we use MDAnalysis to convert a ParmEd structure to an MDAnalysis Universe, select a subset of atoms, and convert it back to ParmEd to simulate with OpenMM.</p>
<p><strong>Last updated:</strong> January 2020</p>
<p><strong>Minimum version of MDAnalysis:</strong> 0.21.0</p>
<p><strong>Packages required:</strong></p>
<ul class="simple">
<li><p>MDAnalysis <a class="reference external" href="#References">[1, 2]</a></p></li>
<li><p>MDAnalysisTests</p></li>
<li><p><a class="reference external" href="http://parmed.github.io/ParmEd/html/index.html">ParmEd</a></p></li>
<li><p><a class="reference external" href="http://openmm.org">OpenMM</a> <a class="reference external" href="#References">[3]</a></p></li>
</ul>
<p><strong>Optional packages for visualisation:</strong></p>
<ul class="simple">
<li><p><a class="reference external" href="http://nglviewer.org/nglview/latest/">nglview</a> <a class="reference external" href="#References">[4]</a></p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">parmed</span> <span class="k">as</span> <span class="nn">pmd</span>
<span class="kn">import</span> <span class="nn">MDAnalysis</span> <span class="k">as</span> <span class="nn">mda</span>
<span class="kn">from</span> <span class="nn">MDAnalysis.tests.datafiles</span> <span class="kn">import</span> <span class="n">PRM7_ala2</span><span class="p">,</span> <span class="n">RST7_ala2</span>
<span class="kn">import</span> <span class="nn">nglview</span> <span class="k">as</span> <span class="nn">nv</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "a191eebb059043ef8d5fffbe76b696c4", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="section" id="Loading-files:-the-difference-between-ParmEd-and-MDAnalysis">
<h2>Loading files: the difference between ParmEd and MDAnalysis<a class="headerlink" href="#Loading-files:-the-difference-between-ParmEd-and-MDAnalysis" title="Permalink to this headline">¶</a></h2>
<p>Both ParmEd and MDAnalysis read a number of file formats. However, while MDAnalysis is typically used to analyse simulations, ParmEd is often used to set them up. This requires ParmEd to read topology parameter information that MDAnalysis typically ignores, such as the equilibrium length and force constants of bonds in the system. For example, the ParmEd structure below.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">pprm</span> <span class="o">=</span> <span class="n">pmd</span><span class="o">.</span><span class="n">load_file</span><span class="p">(</span><span class="n">PRM7_ala2</span><span class="p">,</span> <span class="n">RST7_ala2</span><span class="p">)</span>
<span class="n">pprm</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;AmberParm 3026 atoms; 1003 residues; 3025 bonds; PBC (orthogonal); parametrized&gt;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">pprm</span><span class="o">.</span><span class="n">bonds</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Bond &lt;Atom C [10]; In ALA 0&gt;--&lt;Atom O [11]; In ALA 0&gt;; type=&lt;BondType; k=570.000, req=1.229&gt;&gt;
</pre></div></div>
</div>
<p>When MDAnalysis reads these files in, it does not include that information.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">mprm</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="n">PRM7_ala2</span><span class="p">,</span> <span class="n">RST7_ala2</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;RESTRT&#39;</span><span class="p">)</span>
<span class="n">mprm</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Universe with 3026 atoms&gt;
</pre></div></div>
</div>
<p>The bond type simply shows the atom types involved in the connection.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">mprm</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">bonds</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">type</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(&#39;N3&#39;, &#39;H&#39;)
</pre></div></div>
</div>
<p>If you then convert this Universe to ParmEd, you can see that the resulting Structure is not <code class="docutils literal notranslate"><span class="pre">parametrized</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">mprm_converted</span> <span class="o">=</span> <span class="n">mprm</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">convert_to</span><span class="p">(</span><span class="s1">&#39;PARMED&#39;</span><span class="p">)</span>
<span class="n">mprm_converted</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Structure 3026 atoms; 1003 residues; 3025 bonds; PBC (triclinic); NOT parametrized&gt;
</pre></div></div>
</div>
<p>While the bonds are present, there is no <code class="docutils literal notranslate"><span class="pre">type</span></code> information associated.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">mprm_converted</span><span class="o">.</span><span class="n">bonds</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Bond &lt;Atom N [0]; In ALA 0&gt;--&lt;Atom H1 [1]; In ALA 0&gt;; type=None&gt;
</pre></div></div>
</div>
<p>Therefore, if you wish to use ParmEd functionality that requires parametrization on a MDAnalysis Universe, you need to create that Universe <em>from</em> a ParmEd structure in order to convert it <em>back to</em> something useable in ParmEd.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">mprm_from_parmed</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="n">pprm</span><span class="p">)</span>
<span class="n">mprm_from_parmed</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Universe with 3026 atoms&gt;
</pre></div></div>
</div>
<p>Now the bond type is actually a ParmEd Bond object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">mprm_from_parmed</span><span class="o">.</span><span class="n">bonds</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">type</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Bond &lt;Atom N [0]; In ALA 0&gt;--&lt;Atom H1 [1]; In ALA 0&gt;; type=&lt;BondType; k=434.000, req=1.010&gt;&gt;
</pre></div></div>
</div>
</div>
<div class="section" id="Using-MDAnalysis-to-select-atoms">
<h2>Using MDAnalysis to select atoms<a class="headerlink" href="#Using-MDAnalysis-to-select-atoms" title="Permalink to this headline">¶</a></h2>
<p>One reason we might want to convert a ParmEd structure into MDAnalysis is to use its sophisticated <a class="reference external" href="https://www.mdanalysis.org/UserGuide/selections.html">atom selection syntax</a>. While ParmEd has its <a class="reference external" href="https://parmed.github.io/ParmEd/html/structure.html#structure-manipulation-slicing-combining-replicating-and-splitting">own ways to select atoms</a>, MDAnalysis allows you to select atoms based on geometric distance.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">water</span> <span class="o">=</span> <span class="n">mprm_from_parmed</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;around 5 protein&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">residues</span><span class="o">.</span><span class="n">atoms</span>
<span class="n">protein_shell</span> <span class="o">=</span> <span class="n">mprm_from_parmed</span><span class="o">.</span><span class="n">select_atoms</span><span class="p">(</span><span class="s1">&#39;protein&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="n">water</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">nv</span><span class="o">.</span><span class="n">show_mdanalysis</span><span class="p">(</span><span class="n">protein_shell</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "41a5ce109ebd4cc4a53568e5532fe528", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">prm_protein_shell</span> <span class="o">=</span> <span class="n">protein_shell</span><span class="o">.</span><span class="n">convert_to</span><span class="p">(</span><span class="s1">&#39;PARMED&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">prm_protein_shell</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Structure 155 atoms; 46 residues; 154 bonds; PBC (orthogonal); parametrized&gt;
</pre></div></div>
</div>
</div>
<div class="section" id="Using-ParmEd-and-OpenMM-to-create-a-simulation-system">
<h2>Using ParmEd and OpenMM to create a simulation system<a class="headerlink" href="#Using-ParmEd-and-OpenMM-to-create-a-simulation-system" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">simtk.openmm</span> <span class="k">as</span> <span class="nn">mm</span>
<span class="kn">import</span> <span class="nn">simtk.openmm.app</span> <span class="k">as</span> <span class="nn">app</span>
<span class="kn">from</span> <span class="nn">parmed</span> <span class="kn">import</span> <span class="n">unit</span> <span class="k">as</span> <span class="n">u</span>
<span class="kn">from</span> <span class="nn">parmed.openmm</span> <span class="kn">import</span> <span class="n">StateDataReporter</span><span class="p">,</span> <span class="n">MdcrdReporter</span>
</pre></div>
</div>
</div>
<p>You can create an OpenMM simulation system directly from a ParmEd structure, providing that it is parametrized.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">system</span> <span class="o">=</span> <span class="n">prm_protein_shell</span><span class="o">.</span><span class="n">createSystem</span><span class="p">(</span><span class="n">nonbondedMethod</span><span class="o">=</span><span class="n">app</span><span class="o">.</span><span class="n">NoCutoff</span><span class="p">,</span>
                                        <span class="n">constraints</span><span class="o">=</span><span class="n">app</span><span class="o">.</span><span class="n">HBonds</span><span class="p">,</span>
                                        <span class="n">implicitSolvent</span><span class="o">=</span><span class="n">app</span><span class="o">.</span><span class="n">GBn2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Here we set the integrator to do Langevin dynamics.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">integrator</span> <span class="o">=</span> <span class="n">mm</span><span class="o">.</span><span class="n">LangevinIntegrator</span><span class="p">(</span>
                        <span class="mi">300</span><span class="o">*</span><span class="n">u</span><span class="o">.</span><span class="n">kelvin</span><span class="p">,</span>       <span class="c1"># Temperature of heat bath</span>
                        <span class="mf">1.0</span><span class="o">/</span><span class="n">u</span><span class="o">.</span><span class="n">picoseconds</span><span class="p">,</span>  <span class="c1"># Friction coefficient</span>
                        <span class="mf">2.0</span><span class="o">*</span><span class="n">u</span><span class="o">.</span><span class="n">femtoseconds</span><span class="p">,</span> <span class="c1"># Time step</span>
                        <span class="p">)</span>
</pre></div>
</div>
</div>
<p>We create the Simulation object and set particle positions.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sim</span> <span class="o">=</span> <span class="n">app</span><span class="o">.</span><span class="n">Simulation</span><span class="p">(</span><span class="n">prm_protein_shell</span><span class="o">.</span><span class="n">topology</span><span class="p">,</span> <span class="n">system</span><span class="p">,</span> <span class="n">integrator</span><span class="p">)</span>
<span class="n">sim</span><span class="o">.</span><span class="n">context</span><span class="o">.</span><span class="n">setPositions</span><span class="p">(</span><span class="n">prm_protein_shell</span><span class="o">.</span><span class="n">positions</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>We now minimise the energy.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sim</span><span class="o">.</span><span class="n">minimizeEnergy</span><span class="p">(</span><span class="n">maxIterations</span><span class="o">=</span><span class="mi">500</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The reporter below reports energies and coordinates every 100 steps to stdout, but every 10 steps to the <code class="docutils literal notranslate"><span class="pre">ala2_shell.nc</span></code> file.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sim</span><span class="o">.</span><span class="n">reporters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
        <span class="n">StateDataReporter</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="n">step</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">potentialEnergy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                          <span class="n">kineticEnergy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">temperature</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">volume</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                          <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">sim</span><span class="o">.</span><span class="n">reporters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">MdcrdReporter</span><span class="p">(</span><span class="s1">&#39;ala2_shell.trj&#39;</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">crds</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<p>We can run dynamics for 500 steps (1 picosecond).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sim</span><span class="o">.</span><span class="n">step</span><span class="p">(</span><span class="mi">500</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
#&#34;Step&#34;,&#34;Time (ps)&#34;,&#34;Potential Energy (kilocalorie/mole)&#34;,&#34;Kinetic Energy (kilocalorie/mole)&#34;,&#34;Total Energy (kilocalorie/mole)&#34;,&#34;Temperature (K)&#34;,&#34;Box Volume (angstrom**3)&#34;,&#34;Density (gram/(item*milliliter))&#34;
100,0.20000000000000015,-621.5220464119263,22.782742292451072,-598.7393041194753,72.10507774817847,45325.80641910621,0.034909344712233406
200,0.4000000000000003,-609.8285040481596,37.171151050959445,-572.6573529972002,117.6429378919321,45325.80641910621,0.034909344712233406
300,0.6000000000000004,-605.2386136173518,45.270202024332306,-559.9684115930195,143.2756160228264,45325.80641910621,0.034909344712233406
400,0.8000000000000006,-602.8177469228011,55.762223151457235,-547.0555237713439,176.48180294254348,45325.80641910621,0.034909344712233406
500,1.0000000000000007,-602.1866224008126,68.30392304823927,-533.8826993525734,216.17501610114698,45325.80641910621,0.034909344712233406
</pre></div></div>
</div>
<p>If we write a topology file out from our former <code class="docutils literal notranslate"><span class="pre">protein_shell</span></code> atomgroup, we can load the trajectory in for further analysis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">protein_shell</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;ala2_shell.pdb&#39;</span><span class="p">)</span>
<span class="n">u</span> <span class="o">=</span> <span class="n">mda</span><span class="o">.</span><span class="n">Universe</span><span class="p">(</span><span class="s1">&#39;ala2_shell.pdb&#39;</span><span class="p">,</span> <span class="s1">&#39;ala2_shell.trj&#39;</span><span class="p">)</span>
<span class="n">u</span><span class="o">.</span><span class="n">trajectory</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;TRJReader ala2_shell.trj with 50 frames of 155 atoms&gt;
</pre></div></div>
</div>
<p>[1] R. J. Gowers, M. Linke, J. Barnoud, T. J. E. Reddy, M. N. Melo, S. L. Seyler, D. L. Dotson, J. Domanski, S. Buchoux, I. M. Kenney, and O. Beckstein. <a class="reference external" href="http://conference.scipy.org/proceedings/scipy2016/oliver_beckstein.html">MDAnalysis: A Python package for the rapid analysis of molecular dynamics simulations</a>. In S. Benthall and S. Rostrup, editors, <em>Proceedings of the 15th Python in Science Conference</em>, pages 98-105, Austin, TX, 2016. SciPy, doi:
<a class="reference external" href="https://doi.org/10.25080/majora-629e541a-00e">10.25080/majora-629e541a-00e</a>.</p>
<p>[2] N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein. MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations. <em>J. Comput. Chem</em>. 32 (2011), 2319-2327, <a class="reference external" href="https://dx.doi.org/10.1002/jcc.21787">doi:10.1002/jcc.21787</a>. PMCID:<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3144279/">PMC3144279</a></p>
<p>[3] Peter Eastman, Jason Swails, John D. Chodera, Robert T. McGibbon, Yutong Zhao, Kyle A. Beauchamp, Lee-Ping Wang, Andrew C. Simmonett, Matthew P. Harrigan, Chaya D. Stern, Rafal P. Wiewiora, Bernard R. Brooks, Vijay S. Pande. OpenMM 7: Rapid Development of High Performance Algorithms for Molecular Dynamics. <em>PLoS Comput. Biol.</em> 13:e1005659, 2017.</p>
<p>[4] Hai Nguyen, David A Case, Alexander S Rose. NGLview - Interactive molecular graphics for Jupyter notebooks. <em>Bioinformatics</em>. 34 (2018), 1241–1242, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btx789">doi:10.1093/bioinformatics/btx789</a></p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="analysis/alignment_and_rms/README.html" class="btn btn-neutral float-right" title="Alignments and RMS fitting" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="transformations/center_protein_in_box.html" class="btn btn-neutral float-left" title="Centering a trajectory in the box" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Lily Wang, Richard J Gowers, Oliver Beckstein

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>